staging/src/k8s.io/apiserver/pkg/server/routes/metrics.go

```diff
@@ -20,19 +20,35 @@ import (
 	"io"
 	"net/http"
 
-	apimetrics "k8s.io/apiserver/pkg/endpoints/metrics"
+	"k8s.io/apiserver/pkg/metrics"
 	"k8s.io/apiserver/pkg/server/mux"
+
+	auditmetrics "k8s.io/apiserver/pkg/audit"
+	clientcertmetrics "k8s.io/apiserver/pkg/authentication/request/x509"
+	authmetrics "k8s.io/apiserver/pkg/endpoints/filters"
+	apimetrics "k8s.io/apiserver/pkg/endpoints/metrics"
 	etcdmetrics "k8s.io/apiserver/pkg/storage/etcd/metrics"
+	transformationmetrics "k8s.io/apiserver/pkg/storage/value"
 
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+var allMetrics = metrics.NewStore(
+	prometheus.DefaultRegisterer,
```

Comment:
> My suggestion would be to start refactoring the leaf bits to register metrics onto a metrics registry that is passed down to the registering object, or create and register metrics in the parent object and pass just the pointer to the metrics down. This way we can continue to refactor all the bits until the entire project registers metrics that way. That way will not need any of the sync.Once or reset logic and actually error when metrics are registered multiple times, because that is indeed an error that should not be ignored.

@brancz this is exactly how it's designed - the registry is provided (it's the default prometheus registry, for now, we can make it configurable it in a subsequent PR) with all metrics for this component (`apiserver`). When the `Install` method of the metrics handler is called the metrics are registered.